Add project.publications.official_hca_publication and doi to response (#3502)

danielsotirhos opened the pull request:
https://github.com/DataBiosphere/azul/issues/3502

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

noah-aviel-dove reviewed (approved):
Looks good :+1:

hannes-ucsc commented on test/service/test_repository_projects.py:
```diff
@@ -104,3 +111,46 @@ def assert_file_type_summaries(hit):
+
+@patch_dss_endpoint
+@patch_source_cache
+class TestProjectPublications(WebServiceTestCase):
```
We can't afford to add a full test plus canned bundle for every property we add.

hannes-ucsc commented on test/service/test_response.py:
```diff
@@ -848,6 +848,8 @@ def test_projects_key_search_response(self):
                             "projectTitle": "Single cell transcriptome patterns.",
                             "publications": [
                                 {
+                                    "doi": "10.1016/j.cell.2017.09.004",
```
Can we augment the can used by this test to also have the other fields?

danielsotirhos replied:
The property `official_hca_publication` was added in https://schema.humancellatlas.org/module/project/7.0.0/publication , which requires https://schema.humancellatlas.org/type/project/15.0.0/project or higher.

Instead of augmenting the canned bundle mentioned (which uses https://schema.humancellatlas.org/type/project/9.0.3/project), I've taken `test_project_contact_extraction` from test_hca_indexer.py (which asserts project contact fields) and moved / transformed it into `test_projects_response` in test_response.py to use an existing canned bundle there that already has the desired properties.

amarjandu commented:
#3651 merged, chain shortened

amarjandu commented:
Self note: commit hash before `prsq`: 0df809bd2a7a5e7f8c3449fcb3645be5ced46aa1

hannes-ucsc reviewed (changes requested):
I fixed the call site of assertElasticsearchResultEqual that was added on develop since my push three days ago. Now tests are failing for some other reason, I'm thinking due to removed can.

danielsotirhos commented:
> I fixed the call site of assertElasticsearchResultEqual that was added on develop since my push three days ago. Now tests are failing for some other reason, I'm thinking due to removed can.

Yes, the test was expecting a can removed by e81bec4275065f68e3189e0061359316b3f87a7f . I've added a `!fixup` commit to update the test.

amarjandu commented:
Self note: commit hash before `prsq`: 2d7ec814bbc09d902008c779c31c0a65769e83a6
Ran `make requirements_update` again to update transitive dependencies
